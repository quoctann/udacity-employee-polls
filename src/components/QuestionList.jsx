import { Card } from "react-bootstrap";
import QuestionCard from "./QuestionCard";

const QuestionList = ({ title, questionList, isAnswered }) => {
	return (
		<>
			<Card className="my-4 p-1">
				<Card.Header as="h3" className="text-center">
					{title}
				</Card.Header>
				<Card.Body>
					<div className="d-flex flex-wrap">
						{questionList.map((q) => (
							<QuestionCard key={q.id} question={q} isAnswered={isAnswered} />
						))}
					</div>
				</Card.Body>
			</Card>
		</>
	);
};

export default QuestionList;
