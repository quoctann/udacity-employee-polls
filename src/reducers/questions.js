import {
	ADD_QUESTION,
	ANSWER_QUESTION,
	RECEIVE_QUESTION,
} from "../actions/questions";

export default function questions(state = {}, action) {
	switch (action.type) {
		case RECEIVE_QUESTION:
			return {
				...state,
				...action.questions,
			};
		case ADD_QUESTION:
			return {
				...state,
				[action.question.id]: action.question,
			};
		case ANSWER_QUESTION:
			const question = state[action.questionId];
			const votes = question[action.answer].votes;
			const answer = question[action.answer];
			return {
				...state,
				[action.questionId]: {
					...question,
					[action.answer]: { ...answer, votes: [...votes, action.userId] },
				},
			};
		default:
			return state;
	}
}
