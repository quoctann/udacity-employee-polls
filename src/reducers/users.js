import {
	ADD_USER_QUESTION,
	RECEIVE_USER,
	ADD_USER_ANSWER,
} from "../actions/users";

export default function users(state = {}, action) {
	const { userId, questionId, answer } = action;
	const user = state[userId];

	switch (action.type) {
		case RECEIVE_USER:
			return {
				...state,
				...action.users,
			};
		case ADD_USER_QUESTION:
			return {
				...state,
				[userId]: {
					...user,
					questions: [...user.questions, questionId],
				},
			};
		case ADD_USER_ANSWER:
			const answers = user.answers;
			return {
				...state,
				[userId]: { ...user, answers: { ...answers, [questionId]: answer } },
			};
		default:
			return state;
	}
}
