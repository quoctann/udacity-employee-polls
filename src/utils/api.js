import {
	_getUsers,
	_getQuestions,
	_saveQuestion,
	_saveQuestionAnswer,
	users,
} from "./_DATA.js";

export function getInitialData() {
	return Promise.all([_getUsers(), _getQuestions()]).then(
		([users, questions]) => ({ users, questions })
	);
}

export function saveQuestion(question) {
	return _saveQuestion(question);
}

export function saveQuestionAndAnswer({ authedUser, qid, answer }) {
	return _saveQuestionAnswer({ authedUser, qid, answer });
}

export function login(username, password) {
	return new Promise((resolve, reject) => {
		setTimeout(() => {
			const user = users[username];
			if (!user || user?.password !== password) {
				reject("Username or password incorrect!");
			}

			resolve(user);
		}, 500);
	});
}
