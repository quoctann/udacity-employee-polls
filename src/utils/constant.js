const APP_PATH = {
	ROOT: "/",
	LEADERBOARD: "/leaderboard",
	ADD_QUESTION: "/add",
	QUESTION_DETAIL: "/questions/:question_id",
	ERROR_PAGE: "/error",
};

Object.freeze(APP_PATH);

export { APP_PATH };
