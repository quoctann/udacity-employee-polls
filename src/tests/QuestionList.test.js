import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { create } from "react-test-renderer";
import { createStore } from "redux";
import QuestionList from "../components/QuestionList";
import reducers from "../reducers";
import { questions } from "../utils/_DATA";

describe("Test QuestionList component", () => {
	let testStore;

	beforeAll(() => {
		testStore = createStore(reducers);
	});

	it("Should render correct QuestionList component", async () => {
		let rendering = create(
			<Provider store={testStore}>
				<MemoryRouter>
					<QuestionList questionList={Object.values(questions)} />
				</MemoryRouter>
			</Provider>
		);

		expect(rendering.toJSON()).toMatchSnapshot();
	});
});
