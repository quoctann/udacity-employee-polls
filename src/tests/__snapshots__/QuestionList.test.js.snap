// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test QuestionList component Should render correct QuestionList component 1`] = `
<div
  className="my-4 p-1 card"
>
  <h3
    className="text-center card-header"
  />
  <div
    className="card-body"
  >
    <div
      className="d-flex flex-wrap"
    >
      <div
        className="m-2 card"
        style={
          Object {
            "width": "18rem",
          }
        }
      >
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
            sarahedo
          </div>
          <p
            className="card-text"
          >
            29 Jun 2016, 9:21:12 am
          </p>
          <div
            className="d-grid"
          >
            <button
              className="btn btn-outline-success"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Show
            </button>
          </div>
        </div>
      </div>
      <div
        className="m-2 card"
        style={
          Object {
            "width": "18rem",
          }
        }
      >
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
            mtsamis
          </div>
          <p
            className="card-text"
          >
            14 Jul 2016, 2:02:47 pm
          </p>
          <div
            className="d-grid"
          >
            <button
              className="btn btn-outline-success"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Show
            </button>
          </div>
        </div>
      </div>
      <div
        className="m-2 card"
        style={
          Object {
            "width": "18rem",
          }
        }
      >
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
            sarahedo
          </div>
          <p
            className="card-text"
          >
            04 Mar 2017, 5:22:47 am
          </p>
          <div
            className="d-grid"
          >
            <button
              className="btn btn-outline-success"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Show
            </button>
          </div>
        </div>
      </div>
      <div
        className="m-2 card"
        style={
          Object {
            "width": "18rem",
          }
        }
      >
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
            tylermcginnis
          </div>
          <p
            className="card-text"
          >
            24 Dec 2016, 6:42:47 pm
          </p>
          <div
            className="d-grid"
          >
            <button
              className="btn btn-outline-success"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Show
            </button>
          </div>
        </div>
      </div>
      <div
        className="m-2 card"
        style={
          Object {
            "width": "18rem",
          }
        }
      >
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
            tylermcginnis
          </div>
          <p
            className="card-text"
          >
            15 Mar 2017, 7:09:27 pm
          </p>
          <div
            className="d-grid"
          >
            <button
              className="btn btn-outline-success"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Show
            </button>
          </div>
        </div>
      </div>
      <div
        className="m-2 card"
        style={
          Object {
            "width": "18rem",
          }
        }
      >
        <div
          className="card-body"
        >
          <div
            className="card-title h5"
          >
            mtsamis
          </div>
          <p
            className="card-text"
          >
            01 May 2017, 2:16:07 am
          </p>
          <div
            className="d-grid"
          >
            <button
              className="btn btn-outline-success"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Show
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
